File: src/main/java/org/lanternpowered/server/network/block/BlockEntityProtocol.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.network.block;

import org.lanternpowered.server.block.entity.LanternBlockEntity;
import org.lanternpowered.server.network.vanilla.message.type.play.MessagePlayOutBlockEntity;
import org.spongepowered.api.data.persistence.DataContainer;
import org.spongepowered.api.data.persistence.DataView;

import java.util.function.Supplier;

import org.checkerframework.checker.nullness.qual.Nullable;

public abstract class BlockEntityProtocol<T extends LanternBlockEntity> extends AbstractBlockEntityProtocol<T> {

    /**
     * Constructs a new {@link AbstractBlockEntityProtocol} object.
     *
     * @param blockEntity The block entity
     */
    protected BlockEntityProtocol(T blockEntity) {
        super(blockEntity);
    }

    /**
     * Gets the block entity type that should be used to
     * update the block entity on the client.
     *
     * @return The block entity type
     */
    protected abstract String getType();

    @Override
    protected void init(BlockEntityProtocolUpdateContext context) {
        final DataView dataView = DataContainer.createNew(DataView.SafetyMode.NO_DATA_CLONED);
        populateInitData(dataView);
        context.send(new MessagePlayOutBlockEntity(getType(), this.blockEntity.getLocation().getBlockPosition(), dataView));
    }

    @Override
    protected void update(BlockEntityProtocolUpdateContext context) {
        final LazyDataView lazyDataView = new LazyDataView();
        populateUpdateData(lazyDataView);
        if (lazyDataView.dataView != null) {
            context.send(new MessagePlayOutBlockEntity(getType(), this.blockEntity.getLocation().getBlockPosition(), lazyDataView.dataView));
        }
    }

    static final class LazyDataView implements Supplier<DataView> {

        @Nullable DataView dataView;

        @Override
        public DataView get() {
            if (this.dataView == null) {
                this.dataView = DataContainer.createNew(DataView.SafetyMode.NO_DATA_CLONED);
            }
            return this.dataView;
        }
    }

    @Override
    protected void destroy(BlockEntityProtocolUpdateContext context) {
    }

    /**
     * Populates the {@link DataView} to send to the client.
     *
     * @param dataView The data view
     */
    protected void populateInitData(DataView dataView) {
    }

    /**
     * Populates the {@link DataView} to send to the client.
     *
     * @param dataViewSupplier The data view supplier
     */
    protected void populateUpdateData(Supplier<DataView> dataViewSupplier) {
    }
}

File: src/main/kotlin/org/lanternpowered/server/network/block/BlockEntityProtocol.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.network.block

import org.spongepowered.api.block.entity.BlockEntity

/**
 * Represents the protocol to init, update and destroy a
 * certain [BlockEntity] within a chunk.
 *
 * @property blockEntity The block entity
 * @property T The blockEntity entity type
 */
abstract class BlockEntityProtocol<T : BlockEntity> protected constructor(
        protected val blockEntity: T
) {
    /**
     * The amount of ticks between every update (in ticks).
     */
    var updateInterval = 4

    @JvmField
    var updateTickCounter = 0

    /**
     * Initializes the [BlockEntity] for the
     * given [BlockEntityProtocolUpdateContext].
     *
     * @param context The block entity update context
     */
    abstract fun init(context: BlockEntityProtocolUpdateContext)

    /**
     * Updates the [BlockEntity] for the
     * given [BlockEntityProtocolUpdateContext].
     *
     * @param context The block entity update context
     */
    abstract fun update(context: BlockEntityProtocolUpdateContext)

    /**
     * Destroys the [BlockEntity] for the
     * given [BlockEntityProtocolUpdateContext].
     *
     * @param context The block entity update context
     */
    abstract fun destroy(context: BlockEntityProtocolUpdateContext)
}
